Disable cortex m4 core force-boot on STM32H7

jkrautmacher opened the pull request:
As described in this [issue](https://github.com/zephyrproject-rtos/zephyr/issues/54230) Zephyr currently does not respect the `BCM4` option byte on STM32H7 microcontrollers.

This option byte should disable the boot of the m4 core but the Zephyr kernel on the m7 core force-boots the m4 core. The changes of this merge request disable this force-boot.

jkrautmacher commented:
`grep -R stm32h7_m4_wakeup ~/zephyrproject/zephyr` proved that the removed function is not needed by other parts of the code.

It was in addition tested with basic blinky aswell as a proprietary firmware which needs the `BCM4` option byte. Both times it was tested successfully with `BCM4=0` and `BCM4=1`.

erwango reviewed (changes requested):
Nacking for now as it seems we're not on the same page regarding this piece of code.

`stm32h7_m4_wakeup` does 2 things:
- Enabling HSEM clock which is required for common resource protection in context of dual core execution > You can't remove this in any case if M4 core is started
- Deal with m4 boot depending on BCM4 status:
  - 1: halt M4 core (using M4 core) to wait end of clock initialization which is only done by M7 core
  - 0: start M4 core 

Can you explain your use case ?
As I understand this is different from use cases described here. For instance how do you start M4 core, if not done automatically (BCM4=1), nor by M7 core (BCM4=0)

Note that if you don't want to enable M4 core at all, you can use `CONFIG_STM32H7_DUAL_CORE=n`

jkrautmacher commented:
Thank you for the quick reply.

About the use case: We need to make sure that the m4 is not booting a firmware image which is partially swapped by the MCUboot instance on the m7. Our solution was to simply disable the m4 boot and later start it from the m7 application with `LL_RCC_ForceCM4Boot();`.

About the changes: My first idea was to remove just the `else` clause with `LL_RCC_ForceCM4Boot();` which works too. But since the name `stm32h7_m4_wakeup` implies that it is just needed for waking up the m4 from this m7 init code and it proofed to work I thought it would be cleaner to remove it completely.

The general idea of this merge request is that there is no (to me) known use case where the m4 is unintentionally not started by system reset. Thus a start by the m7 kernel code should not be necessary.

erwango commented:
Thanks for these details. I better get your needs.
No that has to be done is to find a middle ground between your needs and existing code (which is working well so far for other use cases).
I suggest trying to add it as a new Kconfig option.

jkrautmacher commented:
Of course, I can create and force-push a draft for it. 

So the idea is to replace `CONFIG_STM32H7_DUAL_CORE` by that new Kconfig option at the [parts of the code](https://github.com/zephyrproject-rtos/zephyr/pull/54402/files) i changed already?

Would you be ok with the name `CONFIG_STM32H7_WAKE_M4` and a default to no?

erwango commented:
> So the idea is to replace `CONFIG_STM32H7_DUAL_CORE`

Whatever change you propose, please take into account that your need is a new feature and should be backward compatible for others that don't face the same need and should not have impact on their side.
So typically I rather think of a new option that would change behavior only if enabled and have no impact otherwise.

jkrautmacher committed:
soc: arm: Add CONFIG_STM32H7_BOOT_M4_AT_INIT

This new Kconfig option lets the developer configure if the Cortex M4
core should be force-booted during M7 init independent of the BCM4
option byte.

This allows to disable this default behaviour via Kconfig which was not
possible so far.

Signed-off-by: Jan Krautmacher <jan@krautmacher.org>

jkrautmacher commented:
I rewrote the commit history of this merge request. The new single commit adds the `CONFIG_STM32H7_WAKE_M4` Kconfig option which will override BCM4 and is enabled by default for backwards compatibility.

The commit was tested on stm32h747i_disco with the sample/basic/blinky software and works as expected:

|BCM4|CONFIG_STM32H7_WAKE_M4|does M4 blink?|
|----|----------------------|--------------|
|  0 |           n          |      no      |
|  0 |           y          |      yes     |
|  1 |           n          |      yes     |
|  1 |           y          |      yes     |

jkrautmacher commented:
I had to change the commit a bit due to CI errors (another board failed). Now the changes are as minimal as possible and I repeated the testing as described in the table above successfully.

erwango commented on soc/arm/st_stm32/stm32h7/soc_m7.c:
```diff
@@ -37,10 +37,14 @@ static int stm32h7_m4_wakeup(const struct device *arg)
 		 */
 		LL_HSEM_1StepLock(HSEM, CFG_HW_ENTRY_STOP_MODE_SEMID);
-	} else {
+	}
+
+#if defined(CONFIG_STM32H7_WAKE_M4)
+	else {
 		/* CM4 is not started at boot, start it now */
 		LL_RCC_ForceCM4Boot();
 	}
+#endif /* CONFIG_STM32H7_WAKE_M4 */
```
```suggestion
	} else if (IS_ENABLED(CONFIG_STM32H7_WAKE_M4)) {
		/* CM4 is not started at boot, start it now */
		LL_RCC_ForceCM4Boot();
	}
```

jkrautmacher commented on soc/arm/st_stm32/stm32h7/Kconfig.soc:
```diff
@@ -73,3 +73,7 @@ config SOC_STM32H7B3XXQ
 
 endchoice
+
+config STM32H7_WAKE_M4
```
I agree with replacing `WAKE` by `BOOT` and also mentioning `M4` first is totally fine.

The reference to the application confuses me because there are two other options: M4 boot is triggered by hardware with `BCM4=1` or not at all what is at least possible and could be a power saving feature. Furthermore referencing the boot by application refers to what this option is _not doing_ instead of what it _is doing_.

Thus I would propose something like `STM32H7_M4_BOOT_BY_KERNEL` and "Boot M4 core by M7 kernel." with `default y`. I could also think of the order `STM32H7_BOOT_M4_BY_KERNEL` what is closer to natural English grammar but this is not that important.

What do you think?

jkrautmacher replied:
Oh this is good to know! I like the name you suggested. I will force-push the new version in a few minutes.

jkrautmacher replied:
The current version should do the job but the flag definition is still in `Kconfig.soc`. If I move it to `Kconfig.series` as you suggested it becomes an option for SOC selection.

Do you know how to avoid this? I am currently looking trough the Kconfig files and documentation but did not yet find a reason why this behaves like it does.

erwango replied:
> If I move it to `Kconfig.series` as you suggested it becomes an option for SOC selection.

Can you clarify ?

jkrautmacher replied:
I just pushed as a WIP commit what I meant. It does not seem to work. You could maybe point me to the right direction if you have a quick idea. Otherwise I will have a closer look probably next week.

jkrautmacher replied:
Sorry, I am a bit late. This is now solved and pushed.

erwango commented on soc/arm/st_stm32/stm32h7/Kconfig.series:
```diff
@@ -20,3 +20,7 @@ config SOC_SERIES_STM32H7X
 	bool "Dual Core"
 	depends on SOC_SERIES_STM32H7X
+
+config STM32H7_BOOT_M4_AT_INIT
+	bool "Boot M4 core during M7 init independent of option byte BCM4."
+	default y
```
Ok, I see what is happening here.
Please use `soc/arm/st_stm32/stm32h7/Kconfig.defconfig.series` then.
I'll take care same is done for `STM32H7_DUAL_CORE`

jkrautmacher replied:
@erwango This was quite successful. Do you want to move `STM32H7_DUAL_CORE` in a separate merge request or here? If it is just about the move I could also write a commit for that on this branch.

erwango replied:
>  @erwango This was quite successful. Do you want to move STM32H7_DUAL_CORE in a separate merge request or here? If it is just about the move I could also write a commit for that on this branch.

Would be great. Let use another PR, so this does get into the way of current PR

erwango replied:
Looking to compliance check, it seems this is not yet the good solution.
Can you have a try in soc/arm/st_stm32/stm32h7/Kconfig.soc ?

jkrautmacher replied:
Yes, this would probably work since it was the initial attempt which actually worked (and passed the CI I think). I force-pushed such a revision. But after reading the "Boards, SoCs, etc." paragraph from the [Board Porting Guide](https://docs.zephyrproject.org/latest/hardware/porting/board_porting.html) I understand why you initially voted to move this from `Kconfig.soc` to `Kconfig.series` since this is for sure not an option affecting a SOC but instead the whole STM32H7 family.

But placing it in `Kconfig.series` does not work because it is interpreted like a SOC selection option if I remember correctly.
